notebooks/Threshold Optimization Postprocessing for Binary Classification.ipynb

```diff
@@ -0,0 +1,458 @@
+    "sensitive_features_mapped = list(map(lambda x: [x], sensitive_features_test.values))\n",
+    "\n",
+    "FairlearnDashboard(sensitive_features=sensitive_features_mapped,\n",
+    "                   true_y=y_test.values,\n",
+    "                   predicted_ys=predicted_ys,\n",
+    "                   class_names=None,\n",
+    "                   feature_names=X_test.columns.values.tolist(),\n",
+    "                   is_classifier=True)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Postprocessing the model to get a fair model"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The idea behind postprocessing is to alter the output of the fairness-unaware model to achieve fairness. The postprocessing algorithm requires three input arguments:\n",
+    "- the matrix of samples X\n",
+    "- the vector of predictions y from the fairness-unaware model \n",
+    "- the vector of group attribute values A (in our code we refer to it as `sensitive_features`)\n",
+    "\n",
+    "The goal is to make the output fair with respect to constraints. Our postprocessing algorithm uses one of\n",
+    "- Demographic Parity (DP): $P\\ [\\ h(X)=\\hat{y}\\ |\\ A=a] = P\\ [\\ h(X)=\\hat{y}\\ ] \\qquad \\forall a, \\hat{y}$\n",
+    "- Equalized Odds (EO): $P\\ [\\ h(X)=\\hat{y}\\ |\\ A=a, Y=y] = P\\ [\\ h(X)=\\hat{y}\\ |\\ Y=y\\ ] \\qquad \\forall a, \\hat{y}$\n",
+    "\n",
+    "where $h(X)$ is the prediction based on the input $X$, $\\hat{y}$ and $y$ are labels, and $a$ is a sensitive feature value. In our example, we'd expect the postprocessed model with DP to be balanced between races. In this particular scenario it makes more sense to aim at fairness through accuracy like EO. EO does not make the same guarantees. Instead, it ensures that the parity between the subgroups of each race with label 1 in the training set, and parity between the subgroups of each race with label 0 in the training set. Applied to our scenario, this means that the subgroup of each race who reoffended in the past are equally likely to be predicted to reoffend (and therefore also equally likely not to). Similarly, there is parity between subgroups of each race without recidivism, but we have no parity between the groups with different training labels. In mathematical terms at the example of African-American and Caucasian:\n",
+    "\n",
+    "$$\n",
+    "P\\ [\\ \\text{recidivism predicted}\\ |\\ \\text{African-American, recidivism}\\ ] = P\\ [\\ \\text{recidivism predicted}\\ |\\ \\text{Caucasian, recidivism}\\ ], \\text{e.g. } 0.95\\\\\n",
+    "P\\ [\\ \\text{recidivism predicted}\\ |\\ \\text{African-American, no recidivism}\\ ] = P\\ [\\ \\text{recidivism predicted}\\ |\\ \\text{Caucasian, no recidivism}\\ ], \\text{e.g. } 0.15\n",
+    "$$\n",
+    "\n",
+    "but that also means that men (and women) of different subgroup based on training labels don't necessarily have parity:\n",
```

Comment:
I thought we weren't considering gender here?

I'm also confused by this sentence. It appears to be saying that while we want a high true positive rate, we don't want a high false positive rate.... which sort of seems obvious to me.

Reply:
Correct! Somehow must have skipped replacing that sentence. I searched for all occurrences of "men" now and this was the only remaining one. Thanks!

Reply:
That's actually not entirely obvious. Sometimes you're okay with compromising on one of them in order to get the other one up.